packages/ai-jsx/src/batteries/use-tools.tsx

```diff
@@ -219,83 +214,46 @@ export async function* UseTools(props: UseToolsProps, { render }: RenderContext)
 }
 
 /** @hidden */
-export async function* UseToolsFunctionCall(
-  props: UseToolsProps,
-  { render, memo, logger }: ComponentContext
-): RenderableStream {
-  yield AppendOnlyStream;
-
-  const conversation = [memo(props.children)];
-
-  do {
-    const modelResponse = memo(<ChatCompletion functionDefinitions={props.tools}>{conversation}</ChatCompletion>);
-    if (props.showSteps) {
-      yield modelResponse;
-    }
-
-    const renderResult = await render(modelResponse, {
-      stop: (el) => el.tag === AssistantMessage || el.tag == FunctionCall,
-    });
-    let functionCallElement: Element<any> | null = null;
-
-    for (const element of renderResult) {
-      if (isElement(element)) {
-        conversation.push(memo(element));
-
-        if (element.tag === FunctionCall) {
-          // Model has generated a function call.
-          if (functionCallElement) {
-            throw new AIJSXError(
-              `ChatCompletion returned 2 function calls at the same time ${renderResult.join(', ')}`,
-              ErrorCode.ModelOutputCouldNotBeParsedForTool,
-              'runtime'
-            );
+export async function UseToolsFunctionCall(props: UseToolsProps, { render }: ComponentContext) {
+  const converse = (
+    <Converse
+      reply={async (messages, fullConversation) => {
+        const lastMessage = messages[messages.length - 1];
+        switch (lastMessage.type) {
+          case 'functionCall': {
+            const { name, args } = lastMessage.element.props;
+            try {
+              return <FunctionResponse name={name}>{await props.tools[name].func(args)}</FunctionResponse>;
```

Comment:
Not related to PR, but shouldn't `FunctionResponse` have a `failed` flag so that you can tell if it has failed without checking the children?

Reply:
Yeah that seems like a good idea.